File: impl/tools/dsipvec/binding.py
"""WebRTC Media Binding 1.0 (v0.7 companion) — reference semantics for the `media-binding` vectors.

Spec: B§2 (descriptor and SDP; authority rule B§2.1; SDP profile B§2.2), B§3 (roles, DTLS role
from a=setup, codec mapping), B§4 (candidate exchange: carriage, timing/buffering, attribution),
B§5 (renegotiation on the same transport; ICE restart unsupported), B§6.1 (one answer per offer).

Everything here is pure: an SDP mini-parser plus stateless checks and two tiny state machines.
`B§` distinguishes binding sections from Core `§` sections in citations and coverage.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .verdict import Verdict

# B§3.4 — DSIP codec identifiers → SDP encoding names (case-insensitive on the wire).
CODEC_ENCODING = {
    "codec:audio/opus": "opus", "codec:audio/pcmu": "PCMU", "codec:audio/pcma": "PCMA", "codec:audio/g722": "G722",
    "codec:video/h264": "H264", "codec:video/vp8": "VP8", "codec:video/vp9": "VP9", "codec:video/av1": "AV1",
}
DIRECTIONS = ("sendrecv", "sendonly", "recvonly", "inactive")
SECURE_PROTOCOLS = ("UDP/TLS/RTP/SAVPF", "UDP/TLS/RTP/SAVP", "TCP/TLS/RTP/SAVPF", "RTP/SAVPF", "RTP/SAVP")


@dataclass
class Section:
    kind: str
    port: int
    protocol: str
    formats: list[str]
    attrs: list[tuple[str, str | None]] = field(default_factory=list)

    def values(self, name: str) -> list[str]:
        return [v for n, v in self.attrs if n == name and v is not None]

    def has(self, name: str) -> bool:
        return any(n == name for n, _ in self.attrs)

    def direction(self, default: str | None) -> str | None:
        for n, v in self.attrs:
            if n in DIRECTIONS and v is None:
                return n
        return default

    def encodings(self) -> set[str]:
        out = set()
        for v in self.values("rtpmap"):
            parts = v.split(" ", 1)
            if len(parts) == 2:
                out.add(parts[1].split("/")[0].lower())
        return out


@dataclass
class Sdp:
    session_attrs: list[tuple[str, str | None]]
    sections: list[Section]

    def session_value(self, name: str) -> str | None:
        for n, v in self.session_attrs:
            if n == name:
                return v
        return None

    def session_direction(self) -> str | None:
        for n, v in self.session_attrs:
            if n in DIRECTIONS and v is None:
                return n
        return None


def parse_sdp(text: str) -> Sdp | None:
    """Minimal, tolerant SDP parse: `m=` sections and `a=` attributes; everything else ignored."""
    if not isinstance(text, str) or not text.startswith("v=0"):
        return None
    session_attrs: list[tuple[str, str | None]] = []
    sections: list[Section] = []
    for raw in text.replace("\r\n", "\n").split("\n"):
        line = raw.strip()
        if len(line) < 2 or line[1] != "=":
            continue
        key, val = line[0], line[2:]
        if key == "m":
            parts = val.split()
            if len(parts) < 3:
                return None
            try:
                port = int(parts[1].split("/")[0])
            except ValueError:
                return None
            sections.append(Section(parts[0], port, parts[2], parts[3:]))
        elif key == "a":
            name, _, value = val.partition(":")
            attr = (name, value if _ else None)
            (sections[-1].attrs if sections else session_attrs).append(attr)
    return Sdp(session_attrs, sections)


def _attr(sdp: Sdp, sec: Section, name: str) -> str | None:
    vals = sec.values(name)
    return vals[0] if vals else sdp.session_value(name)


def _reject(code: str, reason: str, detail: str) -> Verdict:
    return Verdict.reject(code, reason, detail=detail)


def check_description(payload: dict, *, is_answer: bool, offer_sdp: Sdp | None = None) -> Verdict:
    """B§2.1/B§2.2/B§3.3: the descriptor and its SDP must agree; the SDP must carry the profile.

    Offers fail with `media.unsupported` (or `media.offer-required` without SDP); answers fail
    with `media.failed` (the accepted leg is torn down with `bye`)."""
    bad = "media.failed" if is_answer else "media.unsupported"
    transports = payload.get("transports") or []
    t = transports[0] if transports else {}
    if t.get("id") != "transport:webrtc":
        return Verdict.accept(binding="not-webrtc")
    if not is_answer and t.get("ice") != "trickle":
        return _reject("binding-ice-mode", bad, f"ice={t.get('ice')!r}")
    if is_answer and "ice" in t and t["ice"] != "trickle":
        return _reject("binding-ice-mode", bad, f"ice={t.get('ice')!r}")
    sdp_text = t.get("sdp")
    if not isinstance(sdp_text, str) or not sdp_text:
        return _reject("binding-sdp-missing", "media.failed" if is_answer else "media.offer-required", "transports[0].sdp")
    sdp = parse_sdp(sdp_text)
    if sdp is None:
        return _reject("binding-sdp-invalid", bad, "unparseable SDP")
    media = payload.get("media") or []
    # Answers mirror the offer's section count; rejected sections (port 0) carry no descriptor.
    live = [s for s in sdp.sections if s.port != 0]
    if is_answer and offer_sdp is not None and len(sdp.sections) != len(offer_sdp.sections):
        return _reject("binding-section-count", bad, f"{len(sdp.sections)} sections for an offer of {len(offer_sdp.sections)}")
    if any(s.kind == "application" for s in live):
        return _reject("binding-extra-section", bad, "m=application is outside Core v1.0")
    if len(live) != len(media):
        return _reject("binding-section-count", bad, f"{len(live)} live m= sections for {len(media)} media descriptors")
    for i, (desc, sec) in enumerate(zip(media, live)):
        if sec.kind != desc.get("type"):
            return _reject("binding-kind-mismatch", bad, f"section {i}: m={sec.kind} for type {desc.get('type')}")
        if sec.protocol not in SECURE_PROTOCOLS:
            return _reject("binding-encryption", "media.encryption-required" if not is_answer else bad, f"section {i}: {sec.protocol}")
        direction = sec.direction(sdp.session_direction() or "sendrecv")
        if direction != desc.get("direction"):
            return _reject("binding-direction-mismatch", bad, f"section {i}: a={direction} for direction {desc.get('direction')}")
        enc = sec.encodings()
        for c in desc.get("codecs") or []:
            name = CODEC_ENCODING.get(c.get("id"))
            if name is not None and name.lower() not in enc:
                return _reject("binding-codec-missing", bad, f"section {i}: no rtpmap for {c.get('id')}")
        if not sec.has("rtcp-mux"):
            return _reject("binding-rtcp-mux-missing", bad, f"section {i}")
        fp = _attr(sdp, sec, "fingerprint")
        if fp is None or not fp.lower().startswith("sha-256 "):
            return _reject("binding-fingerprint-missing", bad, f"section {i}: a=fingerprint:sha-256 required")
        if _attr(sdp, sec, "ice-ufrag") is None or _attr(sdp, sec, "ice-pwd") is None:
            return _reject("binding-ice-credentials-missing", bad, f"section {i}")
        setup = _attr(sdp, sec, "setup")
        if is_answer:
            if setup not in ("active", "passive"):
                return _reject("binding-setup-invalid", bad, f"section {i}: answer a=setup:{setup} (must be active or passive)")
        elif setup != "actpass":
            return _reject("binding-setup-invalid", bad, f"section {i}: offer a=setup:{setup} (must be actpass)")
    return Verdict.accept()


def check_offer(payload: dict) -> Verdict:
    """B§2: an `invite`/`update` offer."""
    return check_description(payload, is_answer=False)


def check_answer(offer: dict, answer: dict) -> Verdict:
    """B§2/B§3.1: an `answer` against the offer it selects from; also the ICE-credential rule on re-answers."""
    otr = (offer.get("transports") or [{}])[0]
    osdp = parse_sdp(otr.get("sdp") or "") if isinstance(otr.get("sdp"), str) else None
    return check_description(answer, is_answer=True, offer_sdp=osdp)


def dtls_roles(offer_setup: str, answer_setup: str) -> dict:
    """B§3.3: who is the DTLS client. Offer MUST be actpass; answer MUST be active or passive."""
    if offer_setup != "actpass":
        return Verdict.reject("binding-setup-invalid", "media.unsupported", detail=f"offer a=setup:{offer_setup}").to_expect()
    if answer_setup == "active":
        return {"verdict": "accept", "offerer": "server", "answerer": "client"}
    if answer_setup == "passive":
        return {"verdict": "accept", "offerer": "client", "answerer": "server"}
    return Verdict.reject("binding-setup-invalid", "media.failed", detail=f"answer a=setup:{answer_setup}").to_expect()


def ice_credentials(payload: dict) -> tuple[str | None, str | None]:
    t = (payload.get("transports") or [{}])[0]
    sdp = parse_sdp(t.get("sdp") or "") if isinstance(t.get("sdp"), str) else None
    if sdp is None:
        return None, None
    sec = sdp.sections[0] if sdp.sections else None
    if sec is None:
        return sdp.session_value("ice-ufrag"), sdp.session_value("ice-pwd")
    return _attr(sdp, sec, "ice-ufrag"), _attr(sdp, sec, "ice-pwd")


class CandidateExchange:
    """B§4.2–B§4.4: local candidates are buffered until ACTIVE and sent in signed `info`; the
    end marker is sent exactly once per local description; remote candidates are buffered until
    the remote description is applied, applied in order, ignored after the peer's end marker or
    from any device that is not party to the session, and dropped when the session ends."""

    def __init__(self, ctx: dict):
        self.peer = ctx.get("peer")
        self.active = False
        self.remote_applied = False
        self.local_buf: list[dict] = []
        self.gathering_complete = False
        self.end_sent = False
        self.remote_buf: list[dict] = []
        self.remote_end = False
        self.ended = False

    def step(self, ev: dict) -> list[dict]:
        out: list[dict] = []
        if self.ended:
            return [{"ignore": "ended"}]
        if "local_candidate" in ev:
            if self.active:
                out.append({"send_info": {"candidates": 1, "end_of_candidates": False}})
            else:
                self.local_buf.append(ev["local_candidate"])
                out.append({"buffer": "local", "n": len(self.local_buf)})
        elif "gathering_complete" in ev:
            self.gathering_complete = True
            if self.active and not self.end_sent:
                self.end_sent = True
                out.append({"send_info": {"candidates": 0, "end_of_candidates": True}})
        elif "active" in ev:
            self.active = True
            if self.local_buf or (self.gathering_complete and not self.end_sent):
                end = self.gathering_complete and not self.end_sent
                out.append({"send_info": {"candidates": len(self.local_buf), "end_of_candidates": end}})
                self.local_buf = []
                self.end_sent = self.end_sent or end
        elif "remote_description" in ev:
            self.remote_applied = True
            if self.remote_buf:
                out.append({"apply": len(self.remote_buf)})
                self.remote_buf = []
        elif "remote_info" in ev:
            info = ev["remote_info"]
            if self.peer is not None and info.get("from") != self.peer:
                return [{"ignore": "not-party"}]
            cands = list(info.get("candidates", []))
            if self.remote_end:
                return [{"ignore": "after-end"}]
            if cands:
                if self.remote_applied:
                    out.append({"apply": len(cands)})
                else:
                    self.remote_buf.extend(cands)
                    out.append({"buffer": "remote", "n": len(self.remote_buf)})
            if info.get("end_of_candidates"):
                self.remote_end = True
                out.append({"remote_end": True})
        elif "session_end" in ev:
            self.ended = True
            if self.remote_buf or self.local_buf:
                out.append({"drop_buffered": len(self.remote_buf) + len(self.local_buf)})
                self.remote_buf, self.local_buf = [], []
        return out


class Renegotiation:
    """B§5: a re-offer is a full offer on the same transport (same ICE credentials); the sender
    keeps its current description until the answer applies and rolls back on reject; a remote
    re-offer that changes the ICE credentials is an ICE restart and is refused."""

    def __init__(self, ctx: dict):
        self.ufrag = ctx.get("ufrag")
        self.pending: str | None = None

    def step(self, ev: dict) -> list[dict]:
        if "local_reoffer" in ev:
            u = ev["local_reoffer"].get("ufrag")
            if u != self.ufrag:
                return [{"error": "binding-ice-restart", "detail": "a re-offer MUST keep the ICE credentials"}]
            self.pending = u
            return [{"local_description": "pending"}]
        if "remote_answer" in ev:
            if self.pending is None:
                return [{"ignore": "no-pending-offer"}]
            self.pending = None
            return [{"apply": "answer"}, {"local_description": "current"}]
        if "remote_reject" in ev:
            if self.pending is None:
                return [{"ignore": "no-pending-offer"}]
            self.pending = None
            return [{"rollback": True}, {"local_description": "current"}]
        if "remote_reoffer" in ev:
            u = ev["remote_reoffer"].get("ufrag")
            if u != self.ufrag:
                return [{"reject": {"reason": "media.unsupported", "detail": "ice-restart"}}]
            return [{"ui": "update_offered"}]
        if "answer_update" in ev:
            return [{"apply": "remote-offer+answer"}]
        return [{"ignore": "unknown-event"}]


def one_answer(offer: dict, answers: list[dict]) -> dict:
    """B§6.1 / Core §12.7 rule 4: exactly one answer is applied — the first valid one; earlier
    invalid ones end their leg with `bye media.failed`, later ones with `bye session.already-answered`."""
    applied = None
    legs = []
    for a in answers:
        frm = a.get("from")
        if applied is not None:
            legs.append({"from": frm, "bye": "session.already-answered"})
            continue
        v = check_answer(offer, a)
        if v.ok:
            applied = frm
            legs.append({"from": frm, "applied": True})
        else:
            legs.append({"from": frm, "bye": "media.failed", "code": v.code})
    return {"applied": applied, "legs": legs}


def run(v: dict) -> Any:
    """Harness entry: dispatch on `input.check`."""
    inp, ctx = v["input"], v.get("context", {})
    check = inp["check"]
    if check == "offer":
        return check_offer(inp["payload"]).to_expect()
    if check == "answer":
        return check_answer(inp["offer"], inp["payload"]).to_expect()
    if check == "role":
        return dtls_roles(inp["offer_setup"], inp["answer_setup"])
    if check == "one-answer":
        return one_answer(inp["offer"], inp["answers"])
    if check in ("candidates", "renegotiation"):
        comp = CandidateExchange(ctx) if check == "candidates" else Renegotiation(ctx)
        return {"steps": [{"emit": comp.step(st["event"])} for st in inp["steps"]]}
    raise ValueError(f"unknown media-binding check {check}")

File: impl-ts/src/binding.ts
/**
 * WebRTC Media Binding 1.0: consistency between DSIP media descriptors and the SDP they carry,
 * DTLS roles, trickle-candidate timing, renegotiation, and one answer per offer.
 *
 * Spec: B§2 (transport descriptor), B§2.1 (descriptor/SDP authority), B§2.2 (SDP profile),
 * B§3.3 (DTLS role), B§3.4 (codec mapping), B§4.3–B§4.4 (candidate timing and attribution),
 * B§5 (renegotiation, no ICE restart), B§6.1 (one answer per offer), B§7 (encryption floor),
 * B§8 (failure handling). `B§n` cites `v0.8/dsip-webrtc-media-binding-v0.8.md`.
 */
import type { Json, JsonObject } from "./did.js";
import { reject, type Reject, type Verdict } from "./verdict.js";

const WEBRTC = "transport:webrtc";

/** Spec: B§3.4 — DSIP codec id → the `a=rtpmap` encoding name that must be present. */
const RTPMAP: Record<string, string> = {
  "codec:audio/opus": "opus/48000/2",
  "codec:video/h264": "H264/90000",
  "codec:video/vp8": "VP8/90000",
  "codec:video/av1": "AV1/90000",
};

interface Section {
  kind: string;
  port: number;
  proto: string;
  attributes: string[];
}

interface Sdp {
  session: string[];
  sections: Section[];
}

function parseSdp(text: string): Sdp | null {
  const lines = text.split(/\r?\n/).filter((l) => l.length > 0);
  if (lines[0] !== "v=0" || !lines.every((l) => /^[a-z]=/.test(l))) return null;
  const sdp: Sdp = { session: [], sections: [] };
  for (const line of lines) {
    if (line.startsWith("m=")) {
      const [kind, port, proto] = line.slice(2).split(" ");
      if (kind === undefined || port === undefined || proto === undefined) return null;
      sdp.sections.push({ kind, port: Number(port), proto, attributes: [] });
    } else if (line.startsWith("a=")) {
      (sdp.sections.at(-1)?.attributes ?? sdp.session).push(line.slice(2));
    }
  }
  return sdp.sections.length ? sdp : null;
}

/** An attribute of a section, falling back to the session level. */
function attribute(sdp: Sdp, section: Section, name: string): string | undefined {
  const find = (list: string[]): string | undefined => list.find((a) => a === name || a.startsWith(`${name}:`));
  return find(section.attributes) ?? find(sdp.session);
}

function direction(sdp: Sdp, section: Section): string {
  const names = ["sendrecv", "sendonly", "recvonly", "inactive"];
  return names.find((d) => section.attributes.includes(d)) ?? names.find((d) => sdp.session.includes(d)) ?? "sendrecv";
}

function webrtcTransport(payload: JsonObject): JsonObject | undefined {
  return ((payload["transports"] ?? []) as JsonObject[]).find((t) => t["id"] === WEBRTC);
}

/**
 * Check a descriptor set against its SDP. `as` decides the failure token (B§8): an inconsistent
 * offer is rejected `media.unsupported`; an inconsistent answer fails the accepted leg, `media.failed`.
 */
function checkDescription(payload: JsonObject, as: "offer" | "answer", offered?: Sdp): Reject | Sdp {
  const fail = (code: string, offerReason = "media.unsupported"): Reject => reject(code, as === "offer" ? offerReason : "media.failed");
  const transport = webrtcTransport(payload)!;
  // B§2: `ice` is required on offers; when present it MUST be `trickle`
  if (as === "offer" ? transport["ice"] !== "trickle" : "ice" in transport && transport["ice"] !== "trickle") {
    return fail("binding-ice-mode");
  }
  if (typeof transport["sdp"] !== "string") return fail("binding-sdp-missing", "media.offer-required");
  const sdp = parseSdp(transport["sdp"]);
  if (!sdp) return fail("binding-sdp-invalid");

  // B§2.1: no other m= sections; media[i] is the i-th non-rejected section
  if (sdp.sections.some((s) => s.kind === "application")) return fail("binding-extra-section");
  const media = (payload["media"] ?? []) as JsonObject[];
  const live = sdp.sections.filter((s) => s.port !== 0);
  if (live.length !== media.length) return fail("binding-section-count");
  if (offered && offered.sections.length !== sdp.sections.length) return fail("binding-section-count");

  for (const [i, section] of live.entries()) {
    const d = media[i]!;
    if (section.kind !== d["type"]) return fail("binding-kind-mismatch");
    if (direction(sdp, section) !== ((d["direction"] as string | undefined) ?? "sendrecv")) return fail("binding-direction-mismatch");
    const rtpmaps = section.attributes.filter((a) => a.startsWith("rtpmap:")).map((a) => a.split(" ")[1]?.toLowerCase());
    for (const codec of (d["codecs"] ?? []) as JsonObject[]) {
      const mapped = RTPMAP[codec["id"] as string];
      if (mapped && !rtpmaps.includes(mapped.toLowerCase())) return fail("binding-codec-missing");
    }
  }
  for (const section of live) {
    // B§7: DTLS-SRTP is the floor; plain RTP is never offered or accepted
    if (!section.proto.includes("SAVP")) return fail("binding-encryption", "media.encryption-required");
    // B§2.2
    if (attribute(sdp, section, "rtcp-mux") === undefined) return fail("binding-rtcp-mux-missing");
    if (!attribute(sdp, section, "fingerprint")?.toLowerCase().startsWith("fingerprint:sha-256 ")) return fail("binding-fingerprint-missing");
    if (attribute(sdp, section, "ice-ufrag") === undefined || attribute(sdp, section, "ice-pwd") === undefined) {
      return fail("binding-ice-credentials-missing");
    }
    // B§3.3
    const setup = attribute(sdp, section, "setup")?.slice("setup:".length);
    if (as === "offer" ? setup !== "actpass" : setup !== "active" && setup !== "passive") return fail("binding-setup-invalid");
  }
  return sdp;
}

function isReject(v: Reject | Sdp): v is Reject {
  return "verdict" in v;
}

/** Check an `invite`/`update` body. A payload that does not select `transport:webrtc` is outside this binding. */
export function checkOffer(payload: JsonObject): Verdict {
  if (!webrtcTransport(payload)) return { verdict: "accept", binding: "not-webrtc" };
  const result = checkDescription(payload, "offer");
  return isReject(result) ? result : { verdict: "accept" };
}

/** Check an `answer` body against the offer it answers. Spec: B§3.1 — an answer is an SDP answer, never a counter-offer. */
export function checkAnswer(offer: JsonObject, payload: JsonObject): Verdict {
  if (!webrtcTransport(payload)) return { verdict: "accept", binding: "not-webrtc" };
  const offerSdp = parseSdp((webrtcTransport(offer)?.["sdp"] as string | undefined) ?? "");
  const result = checkDescription(payload, "answer", offerSdp ?? undefined);
  return isReject(result) ? result : { verdict: "accept" };
}

/** DTLS roles from the two `a=setup` values. Spec: B§3.3 */
export function dtlsRoles(offerSetup: string, answerSetup: string): Verdict {
  if (offerSetup !== "actpass") return reject("binding-setup-invalid", "media.unsupported");
  if (answerSetup === "active") return { verdict: "accept", offerer: "server", answerer: "client" };
  if (answerSetup === "passive") return { verdict: "accept", offerer: "client", answerer: "server" };
  return reject("binding-setup-invalid", "media.failed");
}

/** Spec: B§6.1 — apply exactly one answer, the first valid one; release every other leg. */
export function oneAnswer(offer: JsonObject, answers: JsonObject[]): JsonObject {
  let applied: string | null = null;
  const legs = answers.map((answer): Json => {
    const from = answer["from"] as string;
    if (applied !== null) return { from, bye: "session.already-answered" };
    const v = checkAnswer(offer, answer);
    if (v.verdict === "reject") return { from, bye: "media.failed", code: v.code };
    applied = from;
    return { from, applied: true };
  });
  return { applied, legs };
}

/** Trickle-candidate timing for one session. Spec: B§4.3, B§4.4 */
export class Candidates {
  private active = false;
  private ended = false;
  private remoteDescription = false;
  private gatheringComplete = false;
  private endSent = false;
  private remoteEnded = false;
  private local = 0;
  private remote = 0;

  constructor(private readonly peer: string) {}

  /** Apply one event; returns the emissions. */
  step(e: JsonObject): Json[] {
    if (this.ended) return [{ ignore: "ended" }];
    if ("session_end" in e) {
      // B§4.3 rule 3: buffered candidates are dropped, not applied, if the session ends
      this.ended = true;
      const n = this.local + this.remote;
      return n ? [{ drop_buffered: n }] : [];
    }
    if ("local_candidate" in e) {
      // B§4.3 rules 1–2: info is ACTIVE-only, so earlier candidates wait
      if (this.active) return [{ send_info: { candidates: 1, end_of_candidates: false } }];
      this.local += 1;
      return [{ buffer: "local", n: this.local }];
    }
    if ("gathering_complete" in e) {
      if (this.gatheringComplete) return [];
      this.gatheringComplete = true;
      if (!this.active) return [];
      this.endSent = true;
      return [{ send_info: { candidates: 0, end_of_candidates: true } }];
    }
    if ("active" in e) {
      this.active = true;
      if (this.local === 0 && !this.gatheringComplete) return [];
      const out = [{ send_info: { candidates: this.local, end_of_candidates: this.gatheringComplete } }];
      this.endSent = this.gatheringComplete;
      this.local = 0;
      return out;
    }
    if ("remote_description" in e) {
      this.remoteDescription = true;
      const n = this.remote;
      this.remote = 0;
      return n ? [{ apply: n }] : [];
    }
    const info = e["remote_info"] as JsonObject;
    // B§4.4: only the device that is party to the media session
    if (info["from"] !== this.peer) return [{ ignore: "not-party" }];
    if (this.remoteEnded) return [{ ignore: "after-end" }];
    const n = (info["candidates"] as Json[]).length;
    if (!this.remoteDescription) {
      this.remote += n;
      if (info["end_of_candidates"] === true) this.remoteEnded = true;
      return [{ buffer: "remote", n: this.remote }];
    }
    const out: Json[] = n ? [{ apply: n }] : [];
    if (info["end_of_candidates"] === true) {
      this.remoteEnded = true;
      out.push({ remote_end: true });
    }
    return out;
  }
}

/** Local-description handling across a renegotiation. Spec: B§5.2, B§5.4 */
export class Renegotiation {
  private pendingLocal = false;
  private pendingRemote = false;

  constructor(private readonly ufrag: string) {}

  /** Apply one event; returns the emissions. */
  step(e: JsonObject): Json[] {
    if ("local_reoffer" in e) {
      // B§5.4: binding 1.0 has no ICE restart; our own re-offer MUST keep the credentials
      if ((e["local_reoffer"] as JsonObject)["ufrag"] !== this.ufrag) {
        return [{ error: "binding-ice-restart", detail: "a re-offer MUST keep the ICE credentials" }];
      }
      this.pendingLocal = true;
      return [{ local_description: "pending" }];
    }
    if ("remote_answer" in e || "remote_reject" in e) {
      if (!this.pendingLocal) return [{ ignore: "no-pending-offer" }];
      this.pendingLocal = false;
      // B§5.2: on reject, roll back so the media path is exactly the last negotiated one
      return ["remote_answer" in e ? { apply: "answer" } : { rollback: true }, { local_description: "current" }];
    }
    if ("remote_reoffer" in e) {
      if ((e["remote_reoffer"] as JsonObject)["ufrag"] !== this.ufrag) {
        return [{ reject: { reason: "media.unsupported", detail: "ice-restart" } }];
      }
      // B§5.2: the offer is not applied until we decide to answer
      this.pendingRemote = true;
      return [{ ui: "update_offered" }];
    }
    if ("answer_update" in e) {
      if (!this.pendingRemote) return [{ ignore: "no-pending-offer" }];
      this.pendingRemote = false;
      return [{ apply: "remote-offer+answer" }];
    }
    throw new Error(`unknown renegotiation event ${JSON.stringify(e)}`);
  }
}
